internal/pkg/agent/application/upgrade/upgrade.go

```diff
@@ -244,30 +251,51 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
 	// paths.BinaryPath properly derives the binary directory depending on the platform. The path to the binary for macOS is inside of the app bundle.
 	newPath := paths.BinaryPath(filepath.Join(paths.Top(), hashedDir), agentName)
 
-	if err := changeSymlinkInternal(u.log, paths.Top(), symlinkPath, newPath); err != nil {
+	currentVersionedHome, err := filepath.Rel(paths.Top(), paths.Home())
+	if err != nil {
+		return nil, fmt.Errorf("calculating home path relative to top, home: %q top: %q : %w", paths.Home(), paths.Top(), err)
+	}
+
+	if err := ChangeSymlink(u.log, paths.Top(), symlinkPath, newPath); err != nil {
 		u.log.Errorw("Rolling back: changing symlink failed", "error.message", err)
-		rollbackInstall(ctx, u.log, paths.Top(), hashedDir)
+		rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
 		return nil, err
 	}
 
-	currentVersionedHome, err := filepath.Rel(paths.Top(), paths.Home())
-	if err != nil {
-		return nil, fmt.Errorf("calculating home path relative to top, home: %q top: %q : %w", paths.Home(), paths.Top(), err)
+	current := agentInstall{
+		version:       version,
+		hash:          unpackRes.Hash,
+		versionedHome: unpackRes.VersionedHome,
+	}
+
+	previous := agentInstall{
+		version:       release.VersionWithSnapshot(),
+		hash:          release.Commit(),
+		versionedHome: currentVersionedHome,
```

Comment:
It looks like you have swapped the `current` and `previous` names, since the `previous` version is getting the current values of version, hash, and version home.

Reply:
It is correct: we have already rotated the symlink, so in the update we write the `current ` as the version we just upgraded to and the `previous` is the old agent installation. Confusing, I know but the nomenclature of the update marker is based on these definitions

Reply:
Thanks, this would be worth a comment in the code. The `previous.versionedHome.currentVersionedHome` seems suspicious when you read it.

Reply:
22606d7985ef6190e9ab0f89a60305b3136ffe13